### DISCUSSION TITLE
Getting started with Theia and C

### DISCUSSION THREAD

**Discussion by kineticsquid** [eclipse-cdt-cloud/cdt-cloud#38] [Q&A]
Let me preface this with, I haven't done C coding since my undergraduate days many, many years ago.  I'm attempting to get started with C coding in Theia, on OSX. I tried searching for a getting started article, but didn't find anything (for using Theia vs. building on Theia). 

I'm using this simple 'hello world' program. 

```
#include <stdio.h>
 
int main(void)
{
    printf("Hello World");
    int x = 1;
    x = x + 1;
 
    return 0;
}
```

I soon realized I needed gdb. I got that set up with help from this [article](https://ics.uci.edu/~pattis/common/handouts/macmingweclipse/allexperimental/mac-gdb-install.html) and verified with `gdb --version`. 

After poking around a bit, I ended up with the following workflow after creating `hello-world.c`:
- On the bottom margin of the window to the left was a prompt to select a folder. I selected the folder with `hello-world.c`.
<img width="453" alt="Screenshot 2024-02-01 at 8 07 23 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/69b5e71c-5dd9-4c8b-ba21-233761a19e7f">
<img width="666" alt="Screenshot 2024-02-01 at 3 55 53 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/1ec7f649-6b3f-4d65-aef6-db6c399a14c6">

- Then I clicked to compile the program
<img width="568" alt="Screenshot 2024-02-01 at 3 56 25 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/0bbfd384-4cbe-4d77-8db5-3eae6ac7b89e">

- Which resulted in the following artifacts
<img width="409" alt="Screenshot 2024-02-01 at 8 08 56 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/a6fe98b3-1031-451a-a596-27c847e9973b">

- And this output in the `Task:Build` terminal window
<img width="971" alt="Screenshot 2024-02-01 at 8 45 08 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/b957606b-75a0-45e4-b8d9-c70b4257bcdd">

- I set some breakpoints
<img width="331" alt="Screenshot 2024-02-01 at 8 10 51 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/f20c2694-b4d2-4511-8a95-d24bf546321d">

- Then clicked to start debugging
<img width="360" alt="Screenshot 2024-02-01 at 8 11 38 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/9a51cab3-fba9-412c-83d2-e53422f13b8b">

- This resulted in a request for access, which I accepted.
<img width="416" alt="Screenshot 2024-02-01 at 3 58 24 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/cd61dccd-d15f-48b6-af55-f7f3b9bf95dc">

- Then this message in the debug console. I can't find any other error messages not any indication the breakpoints were encountered. 
<img width="832" alt="Screenshot 2024-02-01 at 3 59 03 PM" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/a722a376-e01a-4b75-9183-431b304bfee4">

I'm sure I'm missing something, I just can't seem to determine what that is.

**Comment by planger** [comment]
Thank you for exploring CDT Cloud Blueprint!

Did you configure or create a launch configuration in `launch.json`? If yes, how does it look like? A typical launch configuration for cdt-gdb-adapter would look like the following:
```json
{
  "version": "2.0.0",
  "configurations": [
    {
      "type": "gdb",
      "request": "launch",
      "name": "Debug Example C++ (hello-world)",
      "program": "${workspaceFolder}/hello-world/Example",
      "initCommands": [
        "tbreak main"
      ],
      "preLaunchTask": "Binary build (hello-world)"
    }
  ]
}
```

Also I see that you have installed a number of additional extensions. Which did you install?

If you go to https://try.theia-cloud.io/, login via Github user and select CDT Cloud Blueprint in the dropbox in the bottom, the process should look as follows:
[Screencast 2024-02-02 09:48:10.webm](https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/588090/7918a549-3893-411b-af90-c05a2f77ac8c)


Thank you!

**Comment by kineticsquid** [comment]
@planger Thanks and sorry for the delay in getting back on this. I do have a launch config I created from the dropdown at the top of the debug pane using 'Add configuration...'. I selected 'GDB CDT Local debugging'.  Here's what I ended up with:
```
    {
        "type": "gdb",
        "request": "launch",
        "name": "Hello World",
        "program": "${workspaceFolder}/hello-world/hello-world"
    }
```
Here's what my program looks like:
<img width="370" alt="image" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/ec2e1766-cbdd-4c5a-a70f-aa8590f19ae5">

Based on your launch config, I changed mine as follows:
```
    {
        "type": "gdb",
        "request": "launch",
        "name": "Hello World",
        "program": "${workspaceFolder}/hello-world/hello-world",
        "initCommands": [
            "tbreak main"
          ]
    }
```

Tried debugging again and this time received the following error message: 
```
/Users/johnkellerman/DevelopmentEnvironment/adventure/hello-world/hello-world: No such file or directory.
```

Which made me realize I don't recall how I actually build this program. How does that happen?

I then watched the video at the link above and tried CDT.Cloud Blueprint. When I used the command palette to select the example, this is what I see:
<img width="612" alt="image" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/0685843d-4b91-45ce-9f17-180a4b32eacf">

I tried getting the CMake example in my local Theia deployment, but I don't see it as an option. What extension brings it in? Here's what I currently have:
<img width="379" alt="image" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/2cbf9527-04a8-4b17-af71-45d9fa2be367">

**Reply by planger** [reply]
CDT Cloud Blueprint itself does not dictate how you build your program, you can use any build system.
In the CMake example, which you can generate from the Welcome Page or via the command "Generate CDT Cloud Blueprint Example" -> "CMake Example", we also generate alongside the source code a `tasks.json`:

```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "Binary build (cmake-example)",
      "type": "shell",
      "command": "cmake . && make",
      "group": {
        "kind": "build",
        "isDefault": true
      },
      "options": {
        "cwd": "${workspaceFolder}/cmake-example"
      },
      "problemMatcher": []
    },
    ...
  ]
}
```

This invokes cmake on your project as a shell command (thus cmake must be installed on your system if you run it locally). As this task is marked as default build task, you can also just invoke "Run Build Task", which essentially looks into your task.json and invokes the default build task.

Once the build task is executed, you'll get an executable in `/cmake-example/Example` as a result. This is what is then referenced in `program` of your launch configuration in the `launch.json`:
```json
{
  "version": "2.0.0",
  "configurations": [
    {
      "type": "gdb",
      "request": "launch",
      "name": "Debug Example C++ (cmake-example)",
      "program": "${workspaceFolder}/cmake-example/Example",
      "initCommands": [
        "tbreak main"
      ],
      "preLaunchTask": "Binary build (cmake-example)"
    }
  ]
}
```

In your case, the specified executable `/hello-world/hello-world` seems to be missing. Instead, it looks like you built something in `hello-world/build/Debug/outDebug`, but I'm not sure if this is the executable that resulted from your build.

**Comment by kineticsquid** [comment]
@planger Thanks! I tried the command to generate the CMake example, but I don't seem to have it on my command palette. 

I started fresh with a new folder and my hello-world.c program. I then used the 'Run File' link from the drop down in the upper right corner:
<img width="451" alt="image" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/2f02d5f6-0d61-41d8-b399-c39b24bbb34b">

This resulted in an error and apparently a build executed:
<img width="987" alt="image" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/914c7a5e-3038-40e7-a99a-fdb51c75f7aa">

<img width="523" alt="image" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/d5af46b6-73bf-49c4-bbf1-17671d868b62">

I then used the 'Run File' link again from the drop down to run the program. This seemed to work:
<img width="685" alt="image" src="https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/42de96bf-9baf-42d2-a76a-a2408072c0b0">

Now, with a breakpoint set, I tried debugging, but no luck:
https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/681707/4c59c4b0-f43b-490b-ae70-bc29b6fcff25

It feels like the debugger is unable to connect to running program, like I'm missing a connection definition someplace. BTW, I checked my `.theia` folder and I do not have launch configurations defined.
